### ISSUE TITLE
Use builtins.unsafeGetAttrPos or similar to provide better module system locations

### ISSUE THREAD

**Issue by copumpkin** [NixOS/nixpkgs#15417]
Currently, if I import modules (I'm using the module system for something other than NixOS) that are not top-level files, errors will be reported similarly to:

```
 The option `foo' defined in `<unknown-file>' does not exist.
```

I think `<unknown-file>` could be improved by using `unsafeGetAttrPos` to try a bit harder to figure out where the module is defined. It obviously still won't work in all cases, but it will probably work in more cases.

**Comment by copumpkin** [comment]
cc @nbp

**Comment by nbp** [comment]
Sounds like an interesting idea.  Otherwise, there is `key` top-level attribute of each module to handle cases where we might want to have a cleaner name when the modules are defined in other locations.

**Comment by copumpkin** [comment]
Yeah, perhaps have the displayed location be `unsafeGetAttrPos` unless `key` is defined?

**Comment by copumpkin** [comment]
I've just been working with a bunch of small modules (outside of NixOS) and didn't want to manage a ton of files, and was sad to see the unknown location everywhere.

**Comment by nbp** [comment]
> Yeah, perhaps have the displayed location be unsafeGetAttrPos unless key is defined?

This sounds like a good option.

**Comment by arianvp** [comment]
Unknown location still pops up for me for files outside of Nixpkgs/nixos

**Comment by Infinisil** [comment]
Really we could just use

```nix
(builtins.unsafeGetAttrPos "a" { a = null; }).file
```

as a default for `_file`, this always just returns the current Nix file (or `(string)` if there's no file).

**Comment by roberth** [comment]
This method is used for declarations since
- https://github.com/NixOS/nixpkgs/pull/249243

It could perhaps be applied more broadly, although one caveat is that it wouldn't work well for generated modules such as those produced by `mkRenamedOptionModule`. For those `_file` propagation works rather well. Perhaps `_file` propagation should be opt-in, so that generated modules can select that behavior through a new module attribute.
